docs/spec.md

```diff
@@ -0,0 +1,264 @@
+# Makim Specification
+
+## Overview
+
+Makim uses a YAML-based configuration file (`.makim.yaml`) to define tasks,
+dependencies, and execution environments. This document provides a detailed
+specification of all components available in a Makim configuration file.
+
+## 1. File Structure
+
+A `.makim.yaml` file consists of the following top-level sections:
+
+- `groups`
+- `scheduler`
+- `hosts`
+- `vars`
+- `env`
+- `env-file`
+
+Each section is explained below in detail.
+
+---
+
+## 2. Groups
+
+### Description
+
+Defines collections of related tasks. Each group contains tasks and environment
+variables specific to that group.
+
+### Structure
+
+```yaml
+groups:
+  <group_name>:
+    env:
+      <key>: <value>
+    tasks:
+      <task_name>: <task_definition>
+```
+
+### Example
+
+```yaml
+groups:
+  build:
+    env:
+      BUILD_DIR: dist
+    tasks:
+      clean:
+        help: Clean build artifacts
+        run: rm -rf ${{ env.BUILD_DIR }}
+      compile:
+        help: Compile the project
+        run: echo "Compiling..."
+```
+
+---
+
+## 3. Tasks
+
+### Description
+
+Tasks define executable commands with optional arguments, dependencies, and
+hooks.
+
+### Structure
+
+```yaml
+tasks:
+  <task_name>:
+    help: <description>
+    args: <arguments>
+    env: <environment_variables>
+    hooks: <pre/post-run_hooks>
+    matrix: <parameter_combinations>
+    run: <command>
+```
+
+### Example
+
+```yaml
+tasks:
+  test:
+    help: Run tests
+    args:
+      verbose:
+        type: bool
+        action: store_true
+    run: pytest --verbose=${{ args.verbose }}
+```
+
+---
+
+## 4. Arguments
+
+### Description
+
+Defines arguments that tasks can accept with types, defaults, and help
+descriptions.
+
+### Structure
+
+```yaml
+args:
+  <arg_name>:
+    type: <type>
+    default: <default_value>
+    interactive: <true/false>
+    help: <description>
+```
+
+### Example
+
+```yaml
+args:
+  env:
+    type: str
+    default: "dev"
+    help: Environment setting
+```
+
+---
+
+## 5. Hooks
+
+### Description
+
+Hooks define tasks that run before (`pre-run`) or after (`post-run`) a task
+executes.
+
+### Structure
+
+```yaml
+hooks:
+  pre-run:
+    - task: <task_name>
+  post-run:
+    - task: <task_name>
+```
+
+### Example
+
+```yaml
+hooks:
+  pre-run:
+    - task: clean
+```
+
+---
+
+## 6. Environment Variables
+
+### Description
+
+Defines global, group, or task-specific environment variables.
+
+### Structure
+
+```yaml
+env:
+  <key>: <value>
+```
+
+### Example
+
+```yaml
+env:
+  API_KEY: abc123
+```
+
```

Comment:
add a note at the end, the value should be always a string.
if the developer needs to pass an integer, it should be between quotes.